Add BreakpointEvidence filter based on classifier

TedBrookings opened the pull request:
Added XGBoostEvidenceFilter, an alternative BreakpointEvidence filter
based on XGBoost classifier.
 - Default is still BreakpointDensityFilter. Switch by passing
   "--sv-evidence-filter-type XGBOOST" instead of "DENSITY".
 - Decisions based on evidence overlap or coherence are now scaled based
   on coverage depth (in both filter types).
 - Multiple avenues for supplying saved classifier binary file,
   including built-in resource, local file, and google cloud storage.
 - BreakpointEvidence updated to carry information necessary for
   classifier. Unit tests were correspondingly updated.
 - Data from genome tracts used for some classifier features. From the
   hg38 genome: gaps, centromeres, and umap s100.

Additional changes to convenience scripts
 - Updated sanity_checks.sh to return error signal when exiting
 - Bugfixes to manage_sv_pipeline.sh for linux compatibility
 - Update run_whole_pipeline.sh to detect preemptible workers and
   thus set NUM_EXECUTORS correctly

SHuang-Broad commented on scripts/sv/run_whole_pipeline.sh:
```diff
@@ -43,12 +43,11 @@ SV_ARGS=${*:-${SV_ARGS:-""}}
 # expand any local variables passed as strings (e.g. PROJECT_OUTPUT_DIR)
 eval "SV_ARGS=\"${SV_ARGS}\""
 
-# Choose NUM_EXECUTORS = 2 * NUM_WORKERS
```
I think Chris is currently having a fix for this behavior (i.e. formatting issue of the gcloud commands), see #4767

cwhelan replied:
Yes, I just merged that PR with a fix to `copy_sv_results.sh` that should make this change unnecessary.

TedBrookings replied:
I think a change is still necessary, since run_whole_pipeline.sh executes first and needs to know the number of preemptible + normal workers to set the number of executors. I'm happy to change it to use the same solution you did in your PR.

TedBrookings commented:
Since this touches a lot of files, I'll categorize changes here:
1. Convenience Script Changes
  - bug fixes for running on Linux
    scripts/sv/manage_sv_pipeline.sh
  - detect number of preemptible workers to choose NUM_EXECUTORS correctly
    scripts/sv/run_whole_pipeline.sh
  - allow sanity_checks.sh to run from outside GATK_DIR, exit correctly on error
    scripts/sv/sanity_checks.sh

2. Minor changes to existing utils to support new filter
  - allow construction of IntHistogram.CDF from known cdfFractions and nCounts
    src/main/java/org/broadinstitute/hellbender/tools/spark/utils/IntHistogram.java
  - in constructor, coverage is passed as a float
    src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/ReadMetadata.java
 - add xgboost maven repository for gradle
   build.gradle

3. Significant changes to existing code to support/invoke new filter
 - add arguments for XGBoostEvidenceFilter, changes for scaling density filter by coverage
   src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryArgumentCollection.java
 - replace calls to BreakpointDensityFilter with calls to BreakpointFilterFactory
   src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FindBreakpointEvidenceSpark.java
 - input coverage-scaled thresholds, convert to absolute internally. Allow thresholds to be double instead of int
   src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointDensityFilter.java
 - getter functions added to calculate properties for XGBoostEvidenceFilter. Also fromStringRep() and helper constructors added for testing
   src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java
 - updates to tests reflecting changes to these interfaces
   src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointDensityFilterTest.java
   src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FindBreakpointEvidenceSparkUnitTest.java
   src/test/java/org/broadinstitute/hellbender/tools/spark/sv/integration/FindBreakpointEvidenceSparkIntegrationTest.java
   src/test/java/org/broadinstitute/hellbender/tools/spark/sv/integration/SVIntegrationTestDataProvider.java

4. Added code
 - factory to call appropriate BreakpointEvidence filter
   src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointFilterFactory.java
 - simple helper class to hold feature vectors for classifier
   src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/EvidenceFeatures.java
 - implement classifier-based BreakpointEvidence filter
   src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java
 - unit test for classifier filter
   src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java

5. Added resources
 - Genome tracts
   src/main/resources/large/hg38_centromeres.txt.gz
   src/main/resources/large/hg38_gaps.txt.gz
   src/main/resources/large/hg38_umap_s100.txt.gz
 - Classifier binary file
   src/main/resources/large/sv_evidence_classifier.bin
 - Data used for validation of performance in unit tests
   src/test/resources/sv_classifier_test_data.json
   src/test/resources/sv_features_test_data.json

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
`instanceof` is a potential code smell.  Why not just implement `getMappingQuality` for other types of evidence and set it to 60?

davidbenjamin replied:
Or set is to `Optional.empty` and let this method translate that to 60.

TedBrookings replied:
I've removed "instanceof " in favor of polymorphism with base class returning null by default and classes where values are sensible returning good values.
I've used null instead of Optional.empty because that's the style throughout the project.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Old code?

TedBrookings replied:
Excised

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Using `Optional` as suggested above could reduce the reliance on `instanceof`.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Same about `Optional`.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
`IllegalStateException`

TedBrookings replied:
Changed.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
See above comment about defining a struct-like class for this.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Personally, I would make these `final MutableInt`s and use `increment` below.

TedBrookings replied:
I don't see the advantage. I'd be replacing a non-final primitive with a final class object wrapping a non-final primitive. I did change the python-like += 1 to a proper increment though.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
These underscores are one reason I suggest using a different tuple class in another comment.

TedBrookings replied:
Replaced Tuple2 with ImmutablePair per your later suggestion.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
`tree.put(interval, Collections.singletonList(value))`

TedBrookings replied:
This list needs to be mutable since in principle multiple pieces of BreakpointEvidence can occupy the same location.

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
Can this be `static`?

TedBrookings replied:
Yes. I made this and a few other fields static.

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
There has to already be some natural ordering comparator for contigs somewhere in the GATK or htsjdk.

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
Check in `ArtificialReadUtils` to see if a method like this already exists.

TedBrookings replied:
Unfortunately it doesn't. Changes in response to other comments made it much simpler.

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
It should only very rarely be necessary to write a custom parser in the GATK.  The lowest level one usually needs is to implement a `TableReader`.

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
`cdfFractions = Arrays.stream(cumulativeCounts).map(n -> n / totalObservations).toArray()`

TedBrookings replied:
Unfortunately can't stream this because cdfFraction must be *float* and there's no mapToFloat function.

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
Use `testNG` instead of `JUnit` here and elsewhere.

TedBrookings replied:
This was a subpackage of testNG. I've switched from testng.AssertJUnit to testing.Assert and wrote my own assertArrayEquals()

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
This seems like a good place for a `DataProvider`.

TedBrookings replied:
I'm not sure how to use a DataProvider here. The data is already aggregated and available, I'm just looping through to check if it's consistent.

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
I like "prob" better than "proba"

TedBrookings replied:
Renamed all "Proba" variables to "Proability"

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
`return Arrays.stream(testFeatures).mapToDouble(features -> predictor.predictSingle(features, false,0));`

TedBrookings replied:
Changed to use stream

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
Is there not a Java library that can do a lot of this for you?

TedBrookings replied:
I'm using the jackson JSON library to deal with the hard stuff. However I have to share tabular data from python in the form of pandas DataFrames and numpy arrays with Java. And I'm also passing around heterogenous data (string representations, probabilities, feature matrices). So a lot of this is just niche. Next major PR should be for a java training tool, in which case passing information should be much simpler (serialize / deserialize).

davidbenjamin commented:
@TedBrookings It should go without saying that any time @cwhelan and I disagree, you should listen to him!

davidbenjamin commented:
And also, I agree that if you're working with an API like Spark that is based on iterators then you shouldn't transform them into Streams or a Collections.  Likewise if a problem is inherently online.

davidbenjamin commented:
And while I still think `Tuple2` is inelegant, the fact that it is the convention in Spark overrides the aesthetics.  So I guess I agree with everything Chris said.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryArgumentCollection.java:
```diff
@@ -55,11 +55,11 @@
         @Argument(doc = "Minimum weight of the corroborating read evidence to validate some single piece of evidence.",
                 fullName = "min-evidence-count")
-        public int minEvidenceWeight = 15;
+        public double minEvidenceWeightPerCoverage = 15.0 / 42.855164;
```
Created public static variables to explain these numbers, and commented them.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryArgumentCollection.java:
```diff
@@ -194,8 +202,28 @@
+
+        @Argument(doc = "Path to xgboost classifier model file for evidence filtering",
+                fullName = "sv-evidence-filter-model-file")
+        public String svEvidenceFilterModelFile = "gatk-resources::/large/sv_evidence_classifier.bin";
```
This path is now null by default. Model is loaded from hard-coded (static variable) resource when path is null.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryArgumentCollection.java:
```diff
@@ -194,8 +202,28 @@
+        public String svEvidenceFilterModelFile = "gatk-resources::/large/sv_evidence_classifier.bin";
+
+        @Argument(doc = "Path to single read 100-mer mappability file", fullName = "sv-genome-umap_s100-file")
+        public String svGenomeUmapS100File = "gatk-resources::/large/hg38_umap_s100.txt.gz";
```
Null by default now, and gatk resource removed.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -101,6 +102,102 @@ public String stringRep(final ReadMetadata readMetadata, final int minEvidenceMa
 
+    /**
+     * Returns BreakpointEvidence constructed from string representation. Used to reconstruct BreakpointEvidence for
+     * unit tests. It is intende for stringRep() to be an inverse of this function, but not the other way around. i.e.
```
fixed

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryArgumentCollection.java:
```diff
@@ -104,6 +104,14 @@
 
+        @Argument(doc = "Minimum classified probability for a piece of evidence to pass xgboost evidence filter",
+                fullName = "sv-evidence-filter-threshold-probability")
+        public double svEvidenceFilterThresholdProbability = 0.3;
```
The number was this low because the data set was imbalanced (it favored correct classification of "bad" evidence rather than "good", and therefore was a poor estimator on the data we actually care about). I updated training parameters to correct for that, and the corresponding updated threshold is much closer to 1.0. It should therefore intuitively correspond to high confidence that any given piece of evidence is good.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryArgumentCollection.java:
```diff
@@ -104,6 +104,14 @@
+                fullName = "sv-evidence-filter-threshold-probability")
+        public double svEvidenceFilterThresholdProbability = 0.3;
+
+        @Argument(doc = "Filter method for selecting evidence to group into SV Intervals",
```
Done

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryArgumentCollection.java:
```diff
@@ -194,8 +202,28 @@
+        public String svGenomeUmapS100File = "gatk-resources::/large/hg38_umap_s100.txt.gz";
+
+        @Argument(doc = "Path to file enumerating gaps in the reference genome", fullName = "sv-genome-gaps-file")
+        public String svGenomeGapsFile = "gatk-resources::/large/hg38_gaps.txt.gz";
```
Done

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryArgumentCollection.java:
```diff
@@ -194,8 +202,28 @@
+                fullName = "sv-genome-centromeres-file")
+        public String svGenomeCentromeresFile = "gatk-resources::/large/hg38_centromeres.txt.gz";
+
+        @Argument(doc = "Path to file overriding specification of encoding of categorical variables", optional=true,
```
Moot, this capability is removed due to complexity and lack of need.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -101,6 +102,102 @@ public String stringRep(final ReadMetadata readMetadata, final int minEvidenceMa
+     *      fromStringRep(evidence.stringRep(readMetadata, minEvidenceMapQ), readMetadata) != evidence
+     */
+    @VisibleForTesting
+    static BreakpointEvidence fromStringRep(final String strRep, final ReadMetadata readMetadata) {
```
Moved most new code and all complex logic to test file. Remaining classifiers simply assign values, no complex calculation.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -101,6 +102,102 @@ public String stringRep(final ReadMetadata readMetadata, final int minEvidenceMa
+            final List<StrandedInterval> distalTargets = words[3].isEmpty() ? new ArrayList<>()
+                : Arrays.stream(words[3].split(";")).map(BreakpointEvidence::locationFromStringRep)
+                    .collect(Collectors.toList());
+            if(distalTargets.size() > 1) {
```
Done

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -101,6 +102,102 @@ public String stringRep(final ReadMetadata readMetadata, final int minEvidenceMa
+            if(distalTargets.size() > 1) {
+                throw new IllegalArgumentException("BreakpointEvidence must have 0 or 1 distal targets");
+            }
+            final String[] templateParts = words[4].split("/");
```
The different types of BreakpointEvidence have different properties, and are therefore reconstructed from dumped strings differently. TemplateSizeAnomolies are particularly different from the others, since they are not based on individual reads. Similarly when constructing feature vectors, their features will differ depending on evidence type.

davidbenjamin replied:
I don't see this in the new commits so it's probably moot, but if something like this is really necessary I think it would be safer to move responsibility for parsing into the thing being parsed.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -101,6 +102,102 @@ public String stringRep(final ReadMetadata readMetadata, final int minEvidenceMa
+        final int contig = Integer.parseInt(locationParts[0]);
+        final int start = Integer.parseInt(locationParts[1]);
+        final int end = Integer.parseInt(locationParts[2]);
+        final boolean strand = locationParts.length == 4 && locationParts[3].equals("1");
```
I was trying to get away with only one locationFromStringRep() rather than separate ones for stranded and not. This is confusing problematic for any future development so I have fixed this by having separate functions for stranded locations and unstranded.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -386,19 +495,18 @@ private static SVInterval restOfFragmentInterval( final GATKRead read, final Rea
             this.primaryAlignmentForwardStrand = !read.isReverseStrand();
-            if ( cigar.isEmpty() ) throw new GATKException("Read has no cigar string.");
+            if ( getCigarString() == null || getCigarString().isEmpty() )
+                throw new GATKException("Read has no cigar string.");
```
Fixed

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -410,25 +518,60 @@ public SplitRead( final GATKRead read, final ReadMetadata metadata, final boolea
+                tagSA = null;
+            } else {
+                for(final StrandedInterval distalTarget : distalTargets) {
+                    final String contigName = readMetadata.getContigName(distalTarget.getInterval().getContig());
```
In the SV pipeline contig is stored as an integer code (to save memory). Since I need to occasionally use one to look the other up, I wanted to be clear which is which.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -410,25 +518,60 @@ public SplitRead( final GATKRead read, final ReadMetadata metadata, final boolea
+                    final int referenceLength = distalTarget.getInterval().getLength();
+                    final int pos = distalTarget.getInterval().getEnd() - 1 - UNCERTAINTY;
+                    final int start = strand ? pos - referenceLength: pos;
+                    final int clipLength = 151 - referenceLength;
```
It's not consistently named (although it's also highly possible I've missed some subtleties and there are different strand-related quantities). isForwardStrand seems clearer though, so I changed it. In the code I've added or changed, I've tried to be consistent in using isForwardStrand.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -609,36 +752,51 @@ public int getMapq() {
+                             final List<StrandedInterval> distalTargets) {
+            super(interval, weight, templateName, fragmentOrdinal, validated, forwardStrand, cigarString,
+                    mappingQuality, templateSize, readGroup);
+            // note: passing distalTargets to have uniform testing interface to create BreakpointEvidence
```
Replaced number with call to ReadMetadata::getAvgReadLen() (and moved code to test file).

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -663,6 +821,20 @@ public LargeIndel read( final Kryo kryo, final Input input, final Class<LargeInd
+                             final List<StrandedInterval> distalTargets) {
+            super(interval, weight, templateName, fragmentOrdinal, validated, forwardStrand, cigarString,
+                    mappingQuality, templateSize, readGroup);
+            // note: passing distalTargets to have uniform testing interface to create BreakpointEvidence
```
The constructors now are very simple, and directly assign values with no unused arguments.
In SplitRead and WeirdTemplateSize, more than one value is assigned from one passed object due to requirements that the values be self-consistent, however the identical logic is present in existing non-test code (e.g. in SplitRead(),  this.saMappings = parseSATag(tagSA);)

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -709,6 +881,33 @@ public DiscordantReadPairEvidence(final Kryo kryo, final Input input) {
+                                  final List<StrandedInterval> distalTargets) {
+            super(interval, weight, templateName, fragmentOrdinal, validated, forwardStrand, cigarString,
+                    mappingQuality, templateSize, readGroup);
+            switch(distalTargets.size()) {
```
In principle there could be more than one distal target. The design decision was made throughout the pipeline to keep them as a list, but for now restrict to highest quality distal target if there would otherwise be more than one.

TedBrookings commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/integration/FindBreakpointEvidenceSparkIntegrationTest.java:
```diff
@@ -123,4 +129,48 @@ public void testFindBreakpointRunnableMiniCluster(final FindBreakpointEvidenceSp
+            final List<String> argsToBeModified = Arrays.asList( new ArgumentsBuilder().add(params.getCommandLine()).getArgsArray() );
+            final Path workingDirectory = MiniClusterUtils.getWorkingDir(cluster);
+
+            int idx = 0;
```
Removed

TedBrookings commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/integration/FindBreakpointEvidenceSparkIntegrationTest.java:
```diff
@@ -123,4 +129,48 @@ public void testFindBreakpointRunnableMiniCluster(final FindBreakpointEvidenceSp
+
+            int idx = 0;
+
+            idx = argsToBeModified.indexOf("-I");
```
Done.

TedBrookings commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
I replaced this with a much simpler call to a utility that Steve wrote.

TedBrookings commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
Yes. this is gone now.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Removed

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Fixed keyword order, naming scheme.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
The order matters because the classifier can only look at doubles, and the encoding used by the training routines must be identical to that used by the evaluation. I've come to the conclusion that the override capability is confusing overkill so I've removed it. Just providing a list to specify order seemed like the simplest way to provide the order that had an initializer in java. I'm happy to change it to something simpler/better in response to follow-on comments or a chat.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Yeah, I've had ExternalEvidence on my mind (part of the reason I put the override capability in) but I ignored for now because it hasn't showed up in evidence dumps. Basically I'd need to dump the evidence, train a new classifier, and update the evidence order. I don't think there's a reasonable way to tell a classifier to pretend it knows how to deal with a totally new evidence type, especially because that is the most important feature as ranked by the training algorithm.

cwhelan replied:
I think we should treat `ExternalEvidence` as a first class evidence type (that's on and part of the classifier by default) but to do that we first need to wait until there's a stable gCNV/GATK-SV pipeline and gCNV callset we can use for `ExternalEvidence`. If we want to get fancy we could provide two classifiers, one trained with external evidence and one without. What do you think?

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Created UnscaledOverlapInfo to hold these data.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Yes. This variable was removed by keeping readMetadata instead of unpacking its fields (in response to suggestion by Chris).

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Killed all unnecessary references to "this."

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
I changed this to use FeatureDataSource as suggested by you in a later comment, removing most of this code. Many subsequent criticisms of this design are noted for future reference but not responded to because they're now moot.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Changed to use FeatureDataSource.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Agreed, made that change.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
I made that change. modelFile is optional, null by default. When null, the hard-coded resource file is loaded, otherwise BucketUtils loads the supplied file. Our hg19 and hg38 resource directories each have a copy of the classifier now. I'll have to do some investigation to see how well the hg38-trained classifier performs on hg19-aligned data.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Done.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
The override is gone now. I couldn't figure out how to make a map initializer (that wasn't already basically just appending a list of items). i.e. this seemed like the least verbose way of doing it. I'm open to changing.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Code was removed.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Removed confusing override.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Rendered moot by deletion, but I generally tried to be alert for places where static constants could be extracted.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
This was due to a misunderstanding on my part. I thought the BAM header on the integration test was "bad", not getting it was just hg19. Appropriate genome data is used now, so this crazy hack is removed.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Ok, moot by deletion but won't do again.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Fixed.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
I'd prefer to pass on this one. Basically every piece of evidence executes a loop with this code. Quick checks with JProfiler indicate adding around 2/3 cpu hours added with the stream version. Not huge but I think unnecessary. Ditto for isValidated().

cwhelan replied:
I'm surprised at your result here, @davidbenjamin 's code looks like it should be almost exactly equivalent to the for loop to me. Does your comment mean that it got 66% slower with the stream?

TedBrookings replied:
There's just a tiny overhead, but each piece of evidence starts a for loop (or starts a stream). I profiled a tiny test case, got a roughly 0.1ms cost per BreakpointEvidence, which extrapolated to 40 cpu minutes over the course of a normal BAM file (of course we're running on a cluster, so wall clock increase is greatly less). Because the for loop is so fast, the proportional increase in anyPassesFilter is huge (several-fold), but that doesn't matter so much because it isn't a hot spot. The proportional increase in runtime of the whole pipeline is quite small (IMO just not worth the cost since the for-each loop is quite easy to understand).

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
The classifier can handle missing values. Reasons I didn't set this to NaN:
  1. The features involving sum or mean of mapping quality for overlapping evidence have fairly high importance (overlap_mapping quality and mean_overlap_mapping_quality are more important than num_coherent), and would become problematic when allowing mapping quality to be NaN.
  2. Generally the non-read evidence did seem to have high weight in the Bayesian approaches
  3. The non-read evidence at this point seems like it's inherently well localized (to my understanding it's a change in distribution of read pile-up at a certain point in the reference, so I thought the likelihood of the non-read evidence being in the right *location* was high).

I'll think hard about whether it's possible to replace the "overlap quality" type features with something similar that allows for non-read evidence to have NaN mapping qualities, which I agree is logically more defensible. Right now it would be difficult to switch without substantial redesign of the features.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
I renamed the feature to templateSizeOrReadCount to make it a bit more clear that it's either-or. The alternative is to split into two feature columns and carry around the equivalent of an extra NaN for everything. I think that will make training the classifier (slightly) slower or less effective, so I'd like to avoid that if possible.

cwhelan replied:
Doesn't this possibly allow the classifier to do something silly like cut the data based on this field without regards to evidence type, which doesn't really make any sense? I would think that forcing the classifier to learn that the field means different things for different evidence types would make it less effective, not more, than explicitly telling it that template size anomalies don't have template sizes and read based evidence doesn't have read counts?

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Got rid of all the python-like names.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Done

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Replaced with new class: UnscaledOverlapInfo

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Changed all the accessor methods to private

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Replaced with stream

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Changed return type as suggested.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Changed.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Extracted EvidenceOverlapChecker. Has the additional advantage that it may be useful if alternative filters are developed.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
As I understand it use of iterators is more common in spark projects. In this case I'm making a drop-in replacement for an existing evidence filter and simply continuing the project's use of iterators.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
I originally tried using PairedStrandedIntervalTree, but had trouble with some weird behavior that in retrospect was probably bugs (seeing as you subsequently made a bug fix) but at the time I thought might be deliberate targeting of a particular use case.
However, I think this use case renders it less desirable anyway. I always loop over all overlapping BreakpointEvidence, and also want to know if each piece is coherent (which is an additional O(1) check at that point). I think the PairStrandedIntervalTree would be faster if I *only* wanted the coherent stuff, but as is would not speed things up.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Made ternary

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Changed to ImmutablePair

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
I'm not sure why. Presumably a fossil from previous dev, removed it.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Added break, removed comment.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Removed instanceOf by defining getCigarString for base class to return null.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -101,6 +102,102 @@ public String stringRep(final ReadMetadata readMetadata, final int minEvidenceMa
+            final TemplateFragmentOrdinal fragmentOrdinal;
+            if(templateParts.length <= 1) {
+                fragmentOrdinal = TemplateFragmentOrdinal.UNPAIRED;
+            } else switch (templateParts[1]) {
```
I think @davidbenjamin was missing a parse call to switch the string to an int, and his suggestion is good. This seems to work:

```
            if(templateParts.length <= 1) {
                fragmentOrdinal = TemplateFragmentOrdinal.UNPAIRED;
            } else {
                fragmentOrdinal = TemplateFragmentOrdinal.values()[Integer.parseInt(templateParts[1])];
            }
```

TedBrookings replied:
I'd convinced myself indexing into string enums was impossible, but that works. It compiles but it has the additional problem that I'd have to alter the mapping order.
parseValue -> values index
0 -> 4
1 -> 2
2 -> 3
And that leaves "?" needing to be handled by catching an exception (although I don't *think* any BreakpointEvidence uses that now). If there's a concise way to overload / specify an order for the enum, that could work, otherwise I think the switch statement is easiest. If the switch statement is highly undesirable, another possibility could be to build a static map from String to TemplateFragmentOrdinal, and then I could get all the cases right but just call get() from the map.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
I Forgot to respond to this. I did not explore excluding mismatches. I think it makes a lot of sense to check this out, so I added it to my to-do list before next PR.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
```diff
@@ -28,188 +29,84 @@
-        this.umapS100Mappability = new SVIntervalTree<>();
-        loadGenomeIntervals(this.umapS100Mappability, params.svGenomeUmapS100File, readMetadata);
+        if(params.svGenomeGapsFile == null) {
+            throw new IllegalArgumentException("XGBoostEvidenceFilter requires non-null sv-genome-gaps-file");
```
If you don't have a gaps file, will the classifier not be able to run, or will it just have "0" on the "gap" feature for every piece of evidence? Just wondering if this should be a hard rule.

TedBrookings replied:
It's easy to provide a trivial gaps file (empty file with extension .bed.gz _should_ work), so it seemed like a better idea to ensure they deliberately opted out rather than forgot to specify. Especially since adding gaps helps a lot (3rd place in feature importance). Of course I could set gaps to 0 like you said, and the classifier would do its best assuming you're in a gap-free region.

TedBrookings replied:
Or is there some other clean way to have them specify "no" to the gaps file. i.e. 3 possible states: default (passed nothing), opt out (pass ???), valid file name. I'm not experienced with java arguments, so I don't know the idioms.

cwhelan replied:
Forcing them to make an empty bed file seems like makework. What if you made an additional optional flag `--runWithoutGapsAnnotation` or something like that, and verify that they either provided a value for the file or set the flag?

TedBrookings replied:
That makes sense. I'll make that change and update documentation. I'm thinking I'll need an integration test for that case as well?

TedBrookings replied:
Sorry, one more question. What should happen if they set the flag _and_ pass a valid gaps file? It seems like it's best to use the gaps file in that case, maybe throw a warning?

cwhelan replied:
I think I'd throw an `IllegalArgumentException` there, make them remove one or the other and try not to proceed with nonsensical arguments.

TedBrookings replied:
Ok, I'll do that then.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Since you are summing the mapping qualities here I'd call this field `totalOverlapperMappingQuality` or something like that. Also, why not move computing the mean mapping quality to here, since it's got the sum and the number of overlappers.

TedBrookings replied:
Renamed totalOverlapMappingQuality, and moved calculation of meanOverlapMappingQuality into UnscaledOverlapInfo.

cwhelan replied:
Can you also rename `coherentMappingQuality` to `totalCoherentMappingQuality`? I think I keep looking at those and getting tripped up forgetting that they are actually sums of mapping qualities rather than means or maximums.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
See my comment below about maybe computing the mean in the overlap object instead of here.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
On closer inspection (I think I told you to change this to use `equals` instead of `==`?) it looks like `BreakpointEvidence` and subclasses don't have `equals` and `hashCode` methods. Do you want to add them?  Or, I guess we could go back to using `==`, but that makes me nervous about future bugs resulting from serialization or other weirdness.

TedBrookings replied:
For now I've changed back to ==. Rationale:
    When I made the change I assumed only a horrible person would add a .equals() method to BreakpointEvidence that didn't really ensure equality. That said, I'm really just checking for same object ID, so I'm fine with switching back to ==.
    Thinking about implementing .equals(), I think it might be complicated. I imagine you'd want BreakpointEvidence to check equality of location and evidence type; and ReadEvidence to overload by additionally requiring the same read name?

cwhelan replied:
We probably should just implement `equals` methods on all of the `BreakpointEvidence` implementing classes. They should just check equality of all of the fields in each object since that's the expected behavior. Using `==` makes me pretty nervous; it's generally frowned upon for use on Java objects since if the objects get serialized or cloned they no longer look equal to each other even though all the fields match, which is more often than not a bug. Just use the intellij template to generate them.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Now that I look at this again I'm a bit confused about dividing mapping quality by mean coverage, here and for coherentMappingQuality. That seems somewhat strange to me. What is the reasoning for doing that?

TedBrookings replied:
totalOverlapMappingQuality produces a quantity very similar to numOverlappers, but scaled by mapping quality (e.g. high for overlappers with good mapping quality). Similar to numOverlappers, it needs to be scaled by coverage to ensure proper evaluation of BAMS that have coverage higher or lower than the training data set.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Re-reading this method, I realized that the mappability file contains intervals that overlap -- I think it is all 100mers in the reference that are not unique, right? Doesn't this mean that the value returned by this method can go over 1? Wouldn't you really want the fraction of the evidence that is overlapped by at least one feature from the genome intervals?

TedBrookings replied:
That is a good catch, it didn't even occur to me that the mappability files would contain overlaps. You're correct about the intended goal for this computation. Of course xgboost is quite capable of dealing with the numbers that go over 1 (and maybe overlapping k-mers is better?) but probably fixing will yield better classification power.
    
Of course the same logic error is present in the training python code, so if you want a fix for this PR I'll need to fix the training code and train a new classifier.

cwhelan commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilterUnitTest.java:
These parsing methods are long and tedious (for you to write and others to read). I'm wondering if there are libraries you could use to ease this parsing and cut down on the lines of code. What about GSON, which we include in our dependenies? See here: https://gist.github.com/leoleozhu/5964515 . Could we use that instead to do this parsing work? It's not a huge deal since this is test code but it'd be nice to cut down some of this boilerplate.

TedBrookings replied:
I picked jackson quasi-randomly (it was among the fastest json libraries that was well-liked). The sample GSON code you showed looked simple to read in an array, but the data I'm sending over is nested (multiple objects, tables, matrices, arrays). My recollection is that GSON requires object deserialization which is complex if done directly, but has helper tools. I'd have to study up (or be given instruction) on the best way to do that.
On the other hand, if I get a java training tool running, we can just serialize / deserialize and cut out quite a bit of code.

cwhelan replied:
OK, maybe you want to put in an issue or at least add a TODO comment reminding us to see if that parsing stuff can be simplified.

TedBrookings replied:
Ok. I'm thinking create an issue along the lines of Create Java training tool for XGBoostEvidenceFilter, and have "Eliminate or simplify JSON parsing for test data" as a requirement in the description of that issue. Does that make sense?

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryArgumentCollection.java:
```diff
@@ -55,11 +59,11 @@
 
         @Argument(doc = "Minimum weight of the corroborating read evidence to validate some single piece of evidence.",
                 fullName = "min-evidence-count")
```
We should change the name of the args away from `count` now that they are scaled. I'm blanking a little on what the best names would be.. `min-evidence-coverage-fraction` or something? Also, please update the `doc` string to be a little more descriptive of the new value (and maybe include a small calculation showing that the default value is equivalent to 15 reads at 42X coverage or something?)

TedBrookings replied:
Done.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryArgumentCollection.java:
```diff
@@ -55,11 +59,11 @@
         @Argument(doc = "Minimum weight of the evidence that shares a distal target locus to validate the evidence.",
                 fullName = "min-coherent-evidence-count")
-        public int minCoherentEvidenceWeight = 7;
+        public double minCoherentEvidenceWeightPerCoverage = TRAINING_SET_OPTIMAL_MIN_COHERENCE / TRAINING_SET_MEAN_COVERAGE;
```
Same as above.

TedBrookings commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java:
Done

TedBrookings commented:
Needed two commits because I forgot to check the integration test the first go-round and I had changed parameter args. Sorry!
Github is complaining that there are conflicts, but since local git is fine, I'm assuming that just means I need to rebase before any merge and I should ignore that for now?

Back to you @cwhelan @davidbenjamin

TedBrookings commented:
One oopsie. IntelliJ had no problems with last merge, but Travis and command-line java couldn't handle Class as a type, needed Class<?>. I'm not sure how it's possible for that to happen...

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/BreakpointEvidence.java:
```diff
@@ -51,6 +51,12 @@ protected BreakpointEvidence( final Kryo kryo, final Input input ) {
     public void setValidated( final boolean validated ) { this.validated = validated; }
 
+    // default interfaces to be overridden by derived classes where these values are defined
+    public Integer getMappingQuality() { return null; }
```
Would it work to use `OptionalInt` and `Optional<String>`?  The advantages are that 1) `.isPresent()` is much clearer than ` == null` and 2) the fact that the type is `Optional` makes it self-documenting that it's not necessarily there.

cwhelan replied:
@TedBrookings Even though we talked about this briefly today, I'm looking at these again tonight and wondering why we don't just put them on the appropriate subtypes? Ie. put mapping quality, template size, and cigar string on `ReadEvidence` and read count on `TemplateSizeAnomaly`? Then rather than check for null in `XGBoostEvidenceFilter` check the class type using `instanceof` and get the appropriate values?

TedBrookings replied:
That's how I originally wrote it. It was requested that I use polymorphism and Optional values instead (I used null because that's what was used in prior BreakpointEvidence code). I'm happy to do whatever. Between notes, and conversations, and additional mulling; these are the options I see on the table:
1) use instanceof
2) polymorphism, return null if undefined, check != null
3) polymorphism, return Optional, check Optional::ispresent
4) polymorphism, write checker function e.g. templateSizeIsDefined() but only define getTemplateSize() in subclasses where templateSizeIsDefined() is true
5) polymorphism, write BreakpointEvidence::getFeatures() member function and override. Since overlap and other multi-evidence features are needed, BreakpointEvidence::getFeatures() will be a starting point for computation rather than the end, but it could handle all the NaN logic and return a BreakpointEvidenceFeatures object that's reasonably self-documenting.

I'm totally happy to do any of 1-5 but I think I'd like an official call on this one so I'm not going-round in circles.

tedsharpe replied:
Let me make a suggestion:
I think that the "smelly" aspect of the current code -- the part that seems not quite properly object oriented -- is testing for a particular concrete class:  ReadEvidence.
This should be an interface, rather than a concrete class.  Asking a piece of evidence, "Are you the kind of evidence that's associated with a particular read?" is OK in my book.  (And then you downcast to the interface that allows you to interrogate properties of the associated read.)
So I'd recommend splitting the current ReadEvidence class into an interface and a concrete implementation.  I should have done this to begin with.  Sorry I took a shortcut.

davidbenjamin replied:
@tedsharpe's suggestion sounds good to me.  I'm dropping out of this one so as to avoid too many cooks spoiling broth.

cwhelan replied:
This is a pretty minor debate over a few lines of code -- OO purists would tell us to implement a Visitor pattern but that's way too much work and would too tightly couple the features needed for the classifier to the classifier IMO. @tedsharpe 's proposal sounds like a good compromise. I can also see the use of `Optionals` in this case. @TedBrookings  I'm usually fine with checking `nulls`, as I said yesterday, but in this case where null is being used for a subclass-specific data field (and therefore sort of overriding the inheritance hierarchy of the class) it does seem pretty ugly.

cwhelan commented:
I think I'm done looking over this current round. It looks good to me apart from the relatively minor things I've noted and we've talked about.

TedBrookings commented:
I removed the unused code paths and make a new branch (tb_sv_save_classifier_feature_strategies) that I can cherry-pick to switch strategies. I've rebased on master, so should be good to merge once travis finishes and @davidbenjamin approves.

davidbenjamin commented:
:thumbsup:!
